IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,666 @@
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        return nil;
+    }
+
+    NSArray *items = CFBridgingRelease(cfItems);
+
+    query = [self primaryAttributesForKey:key];
+    [query addEntriesFromDictionary:[self secondaryAttributesForKey:key]];
+    // Note: For efficiency, use kSecUseItemList to query the items returned above rather actually querying
+    // the keychain again. With this second query we can set a specific kSecMatchLimit which lets us get the data
+    // objects.
+    query[(id)kSecUseItemList] = items;
+    query[(id)kSecMatchLimit] = @(items.count + 1); // always set a limit > 1 so we consistently get an NSArray result
+    query[(id)kSecReturnAttributes] = @YES;
+    query[(id)kSecReturnData] = @YES;
+
+    CFTypeRef cfItemDicts = nil;
+    status = SecItemCopyMatching((CFDictionaryRef)query, &cfItemDicts);
+    NSArray *itemDicts = CFBridgingRelease(cfItemDicts);
+
+    NSMutableArray<MSIDAccountCacheItem *> *accountList = [NSMutableArray new];
+    for (NSDictionary *dict in itemDicts)
+    {
+        NSData *jsonData = dict[(id)kSecValueData];
+        if (jsonData)
+        {
+            MSIDAccountCacheItem *account = (MSIDAccountCacheItem *)[serializer deserializeAccountCacheItem:jsonData];
+            if (account != nil)
+            {
+                [accountList addObject:account];
+            }
+            else
+            {
+                MSID_LOG_WARN(context, @"Failed to deserialize account");
+            }
+        }
+    }
+    return accountList;
+}
+
+// Remove one or more accounts from the keychain that match the key.
+//
+// Errors:
+// * MSIDKeychainErrorDomain/OSStatus: Apple status codes from SecItemDelete()
+//
+- (BOOL)removeItemsWithAccountKey:(MSIDCacheKey *)key
+                          context:(id<MSIDRequestContext>)context
+                            error:(NSError **)error
+{
+    MSID_TRACE;
+    MSID_LOG_VERBOSE(context, @"Remove keychain items, key info (account: %@ service: %@, keychainGroup: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service), [self keychainGroupLoggingName]);
+    MSID_LOG_VERBOSE_PII(context, @"Remove keychain items, key info (account: %@ service: %@, keychainGroup: %@)", key.account, key.service, self.keychainGroup);
+
+    if (!key)
+    {
+        [self createError:@"Key is nil."
+                   domain:MSIDErrorDomain errorCode:MSIDErrorInvalidDeveloperParameter error:error context:context];
+        return NO;
+    }
+
+    NSMutableDictionary *query = [self primaryAttributesForKey:key];
+    [query addEntriesFromDictionary:[self secondaryAttributesForKey:key]];
+    query[(id)kSecMatchLimit] = (id)kSecMatchLimitAll;
+
+    MSID_LOG_INFO(context, @"Trying to delete keychain items...");
+    OSStatus status = SecItemDelete((CFDictionaryRef)query);
+    MSID_LOG_INFO(context, @"Keychain delete status: %d", (int)status);
+
+    if (status != errSecSuccess && status != errSecItemNotFound)
```

Comment:
why do we return an error when it's not found?

Reply:
That's how the iOS implementation is written, as well as the in-memory Mac cache.

Reply:
Update: this makes sense that we don't fail when we don't find an item